awscli/customizations/autoprompt/history.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import os
+
+from prompt_toolkit.completion import Completion, Completer
+from prompt_toolkit.history import FileHistory
+
+from awscli.autocomplete.completer import CompletionResult
+from awscli.autocomplete.filters import fuzzy_filter
+
+
+class HistoryDriver(FileHistory):
+    def __init__(self, filename, max_commands=500):
+        # keep only last max_commands commands in file, each command
+        # takes three lines
+        super().__init__(filename)
+        self._max_lines = (max_commands-1) * 3
+
+    def load_history_strings(self):
+        try:
+            return super().load_history_strings()
+        except Exception:
+            return []
+
+    def store_string(self, string):
+        try:
+            if os.path.exists(self.filename):
+                with open(self.filename, 'r') as f:
+                    lines = f.readlines()
+                if len(lines) > self._max_lines:
+                    with open(self.filename, 'w') as f:
+                        f.writelines(lines[-self._max_lines:])
+            super().store_string(string)
+        except Exception:
+            pass
+
+
+class HistoryCompleter(Completer):
+
+    def __init__(self, buffer):
+        self.buffer = buffer
+        self.working_lines = buffer.history.get_strings()
```

Comment:
We should probably just make this a lazy load as we are loading it from a file:
```python
    def __init__(self, buffer):
        self.buffer = buffer
        self._working_lines = None

   @property
    def working_lines(self):
        if self._working_lines is None:
            self._working_lines = self.buffer.history.get_strings()
        return self._working_lines
```

Reply:
Prompt toolkit starts loading strings in background on buffer creation so this laziness won't change anything